lib/src/listenable_provider.dart

```diff
@@ -0,0 +1,232 @@
+
+  @override
+  ListenableProvider<T> cloneWithChild(Widget child) {
+    return ListenableProvider<T>._(
+      key: key,
+      builder: builder,
+      listenable: listenable,
+      dispose: dispose,
+      child: child,
+    );
+  }
+}
+
+class _ListenableProviderState<T extends Listenable>
+    extends State<ListenableProvider<T>> {
+  static bool didChangeBetweenDefaultAndBuilderConstructor(
+    ListenableProvider oldWidget,
+    ListenableProvider widget,
+  ) =>
+      isBuilderConstructor(oldWidget) != isBuilderConstructor(widget);
+
+  static bool isBuilderConstructor(ListenableProvider provider) =>
+      provider.builder != null;
+
+  /// The number of time [Listenable] called its listeners.
+  ///
+  /// It is used to differentiate external rebuilds from rebuilds caused by the listenable emitting an event.
+  /// This allows [InheritedWidget.updateShouldNotify] to return true only in the latter scenario.
+  int _buildCount;
+  UpdateShouldNotify<T> updateShouldNotify;
+  T listenable;
+
+  @override
+  void initState() {
+    super.initState();
+    startListening();
+  }
+
+  void listener() {
+    setState(() {
+      _buildCount ??= 0;
+      _buildCount++;
+    });
+  }
+
+  void startListening() {
+    listenable = widget.listenable ??
+        (widget.builder != null ? widget.builder(context) : null);
+    updateShouldNotify = createUpdateShouldNotify();
+    listenable?.addListener(listener);
+  }
+
+  void stopListening(ListenableProvider<T> widget) {
+    listenable?.removeListener(listener);
+    if (widget.dispose != null) {
+      // if we have a dispose, then we're using builder constructor so we can safely call the method
+      widget.dispose(context, listenable);
+    }
+    listenable = null;
+  }
+
+  UpdateShouldNotify<T> createUpdateShouldNotify() {
+    final capturedBuildCount = _buildCount;
+    return updateShouldNotify = (_, __) => _buildCount != capturedBuildCount;
+  }
+
+  @override
+  void didUpdateWidget(ListenableProvider<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (didChangeBetweenDefaultAndBuilderConstructor(oldWidget, widget) ||
+        widget.listenable != oldWidget.listenable) {
+      stopListening(oldWidget);
+      startListening();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Provider<T>(
+      value: listenable,
+      child: widget.child,
+      updateShouldNotify: updateShouldNotify,
+    );
+  }
+
+  @override
+  void dispose() {
+    stopListening(widget);
+    super.dispose();
+  }
+}
+
+class ChangeNotifierProvider<T extends ChangeNotifier>
+    extends ListenableProvider<T> implements SingleChildClonableWidget {
+  static void _disposer(BuildContext context, ChangeNotifier notifier) =>
+      notifier?.dispose();
+
+  const ChangeNotifierProvider({
+    Key key,
+    T notifier,
+    Widget child,
+  }) : super(key: key, listenable: notifier, child: child);
+
+  const ChangeNotifierProvider.builder({
```

Comment:
More feedback: I prefer the old `stateful` constructor instead of `builder` for API consistency as well (thinking along the lines of "Ah, this like a StatefulProvider but for ChangeNotifiers!")

Reply:
I don't like this one. Because then peoples would request `Provider.stateful`, and ultimately `Provider.whatever` instead of `WhateverProvider` when it doesn't necessarily make sense.

Reply:
I, for one, slightly prefer "builder" over "stateful". My main reason is accessibility for newcomers. Stateful seems like implementation-leaking. What the user cares about is that they'll build a value to be provided. See a top-level comment below for a more general discussion about builder/stateful.